### PR TITLE
update to 4.23.4

### PR DESCRIPTION
**Pull request by M-Waszkiewicz-Anaconda**
dependency for snowflake dbt
- [upstream](https://github.com/protocolbuffers/protobuf/tree/v23.4)
- remove patches from 3.x
- add patch to use C++17 only
- update home. dev, doc urls

### CODE REVIEW

File: recipe/meta.yaml
Comment (skupr-anaconda) [inline recipe/meta.yaml]:
```suggestion
    - patch     # [not win]
```

File: recipe/meta.yaml
Comment (skupr-anaconda) [inline recipe/meta.yaml:14]: Should we update build.sh/bld.bat to run:
`pip install . -vv --install-option="--cpp_implementation" -- no-deps --no-build isolation` like conda-forge did?
See https://github.com/conda-forge/protobuf-feedstock/blob/main/recipe/meta.yaml#L30C23-L30C80

And I wonder if the env variables are needed too https://github.com/conda-forge/protobuf-feedstock/blob/20df32ca656756cbb2401ec50dcbd70ee2d0fe6b/recipe/meta.yaml#L25-L29

File: recipe/meta.yaml
Comment (skupr-anaconda) [inline recipe/meta.yaml:26]: Missing `setuptools` and `wheel`

File: recipe/meta.yaml
Comment (skupr-anaconda) [inline recipe/meta.yaml:29]: Do we need `abseil-cpp`  https://github.com/protocolbuffers/protobuf/blob/e45a6d50682d5d4b204e7d76254b1585b43c6815/src/README.md?plain=1#L22? The conda-forge recipe has `libabseil` in `host` https://github.com/conda-forge/protobuf-feedstock/blob/20df32ca656756cbb2401ec50dcbd70ee2d0fe6b/recipe/meta.yaml#L43

File: recipe/meta.yaml
Comment (skupr-anaconda) [inline recipe/meta.yaml:23]: FYI, in the future, the `upd` C library will be the main backend instead of C++ https://github.com/protocolbuffers/protobuf/tree/v23.4/python#implementation-backends but conda-forge still uses C/C++

File: recipe/meta.yaml
Comment (M-Waszkiewicz-Anaconda) [reply recipe/meta.yaml:14]: pip install done, about env variables 2 of them are for cross compilation, the last one for windows and they are in the build section so if build passes I suppose we do need them, I would prefer to not adding them until we know what they for.

File: recipe/meta.yaml
Comment (M-Waszkiewicz-Anaconda) [reply recipe/meta.yaml:29]: it is just name difference, we have abseil-cpp and CF has libabseil, maybe we should another PR for abseil and make alias  for package name as we have in boost also similar situation ?

Comment (skupr-anaconda) [reply recipe/meta.yaml:29]: IMO, there is a name difference only in our abseil recipe and cf one.  I think it differs from the boost situation

File: recipe/meta.yaml
Comment (skupr-anaconda) [reply recipe/meta.yaml:14]: conda-forge logs say:
`DEPRECATION: --install-option is deprecated because it forces pip to use the 'setup.py install' command which is itself deprecated. pip 23.1 will enforce this behaviour change. A possible replacement is to use --config-settings. Discussion can be found at https://github.com/pypa/pip/issues/11358`
See https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=760479&view=logs&j=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e&t=841356e0-85bb-57d8-dbbc-852e683d1642&l=214

Ah, that's why conda-forge uses `pip <23.1` in `host` https://github.com/conda-forge/protobuf-feedstock/blob/20df32ca656756cbb2401ec50dcbd70ee2d0fe6b/recipe/meta.yaml#L41

If you apply the same pinning it will work. But a comment will clarify why we do this

File: recipe/meta.yaml
Comment (M-Waszkiewicz-Anaconda) [reply recipe/meta.yaml:29]: about missing abseil in the host, it is indirect dependency pulled by libprotobuf, eg:

```
The following NEW packages will be INSTALLED:
    ...
    abseil-cpp:       20230802.0-h6a678d5_2  local
    ...
```

File: recipe/meta.yaml
Comment (M-Waszkiewicz-Anaconda) [reply recipe/meta.yaml:14]: running
` "%PYTHON%" -m pip install . -vv --install-option="--cpp_implementation" --no-deps --no-build-isolation`
on windows, gave error:
`DEPRECATION: protobuf is being installed using the legacy 'setup.py install' method, because the '--no-binary' option was enabled for it and this currently disables local wheel building for projects that don't have a 'pyproject.toml' file. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/11451`

**Review by skupr-anaconda — APPROVED** [review]
LGTM but I would wait for reviews from the Marquee pod as protobuf is a dependency for ray and tensorflow packages

File: recipe/meta.yaml
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml:31]: I would add it back like this `- libprotobuf  # pin handled through libprotobuf's run_exports` so that it's more explicit.

File: recipe/patches/0004-use-C-17-everywhere.patch
Comment (JeanChristopheMorinPerso) [inline recipe/patches/0004-use-C-17-everywhere.patch:7]: Can you add a link to the original patch please?

File: recipe/build.sh
Comment (JeanChristopheMorinPerso) [inline recipe/build.sh]: `--install-option` is deprecated.
Pip complains:
>2023-08-29T10:16:49.978234+00:00  | INFO     |   DEPRECATION: --install-option is deprecated because it forces pip to use the 'setup.py install' command which is itself deprecated. pip 23.1 will enforce this behaviour change. A possible replacement is to use --config-settings. Discussion can be found at https://github.com/pypa/pip/issues/11358

I think you can do `pip install . --config-settings="--build-option=--cpp_implementation"`.

**Review by JeanChristopheMorinPerso — CHANGES_REQUESTED** [review]

File: recipe/build.sh
Comment (M-Waszkiewicz-Anaconda) [reply recipe/build.sh]: will try this

Comment (M-Waszkiewicz-Anaconda) [reply recipe/build.sh]: it does work only for windows, I would leave the one for windows and get back to previous version for unix, right @JeanChristopheMorinPerso  ?

It fails with
` python -c 'from google.protobuf.internal import api_implementation; assert api_implementation.Type() == '\''cpp'\'''`

Comment (M-Waszkiewicz-Anaconda) [reply recipe/build.sh]: actually I will revert both as windows we do not have a tests for it

File: recipe/patches/0004-use-C-17-everywhere.patch
Comment (M-Waszkiewicz-Anaconda) [reply recipe/patches/0004-use-C-17-everywhere.patch:7]: put link in meta.yaml

File: recipe/meta.yaml
Comment (M-Waszkiewicz-Anaconda) [reply recipe/meta.yaml:31]: done

File: recipe/patches/0004-use-C-17-everywhere.patch
Comment (JeanChristopheMorinPerso) [reply recipe/patches/0004-use-C-17-everywhere.patch:7]: I would patch it in the patch directly.

File: recipe/build.sh
Comment (M-Waszkiewicz-Anaconda) [reply recipe/build.sh]: looks like windows build does not require --build-option

File: recipe/meta.yaml
Comment (JeanChristopheMorinPerso) [inline recipe/meta.yaml]: Keep `- libprotobuf {{ version }}` in the host section and would move this line in the run section.

**Review by JeanChristopheMorinPerso — APPROVED** [review]
Great work @M-Waszkiewicz-Anaconda !

**Pull request closed without merging**
